### PR TITLE
Support adding firewall for shoot apiserver and shoot VPN

### PR DESCRIPTION
**Pull request by aylei**
Related change in aws provider: https://github.com/pingcap/gardener-extension-provider-aws/commit/dfed5a9c89709d8a01f8d740eea394c35f4084b8

Tested in dev:

Shoot apiserver firewall:
```
apiVersion: v1
kind: Service
metadata
  name: kube-apiserver
spec:
  loadBalancerSourceRanges:
  # Mocked PingCAP or Central
  - 101.96.***.190/32
  # Shoot outbounds
  - 54.184.***.54/32
  - 50.112.***.142/32
  - 52.25.***.173/32
```

VPN shoot firewall:

```
apiVersion: v1
kind: Service
metadata:
  name: vpn-shoot
  namespace: kube-system
spec:
  loadBalancerSourceRanges:
  # Seed outbounds
  - 54.188.***.195/32
  - 54.69.***.57/32
  - 44.231.***.187/32
```

### CODE REVIEW

File: pkg/operation/botanist/addons.go [hunk 1]
```diff
@@ -378,6 +379,24 @@ func (b *Botanist) generateCoreAddonsChart() (*chartrenderer.RenderedChart, erro
 	}
 	shootInfo["extensions"] = strings.Join(extensions, ",")
 
+	cm, err := b.K8sSeedClient.Kubernetes().CoreV1().ConfigMaps("kube-system").Get("vpc-outbound-ips", metav1.GetOptions{})
```
Comment (aylei) [inline pkg/operation/botanist/addons.go:382]: Configure the firewall of VPN shoot if the Seed declares its outbound ips

File: pkg/operation/botanist/controlplane.go [hunk 2]
```diff
@@ -670,7 +670,17 @@ func (b *Botanist) DeployNetworkPolicies(ctx context.Context) error {
 // DeployKubeAPIServerService deploys kube-apiserver service.
 func (b *Botanist) DeployKubeAPIServerService(ctx context.Context) error {
 	const name = "kube-apiserver-service"
-	return b.ChartApplierSeed.Apply(ctx, filepath.Join(chartPathControlPlane, name), b.Shoot.SeedNamespace, name)
+	return b.ChartApplierSeed.Apply(ctx, filepath.Join(chartPathControlPlane, name), b.Shoot.SeedNamespace, name, kubernetes.MergeFuncs{
```
Comment (aylei) [inline pkg/operation/botanist/controlplane.go:673]: Do not touch the `loadBalancerSourceRanges` field when apply kube-apiserver svc, we will manage this field in another place

File: pkg/operation/botanist/infrastructure.go [hunk 3]
```diff
@@ -132,6 +134,54 @@ func (b *Botanist) WaitUntilInfrastructureReady(ctx context.Context) error {
 	)
 }
 
+// DeployApiServerFirewall add the shoot worker nodes as a trust source of shoot apiserver
+func (b *Botanist) DeployApiServerFirewall(ctx context.Context) error {
```
Comment (aylei) [inline pkg/operation/botanist/infrastructure.go:138]: Where the firewall of shoot-apiserver added

File: pkg/operation/botanist/infrastructure.go [hunk 4]
```diff
@@ -132,6 +134,54 @@ func (b *Botanist) WaitUntilInfrastructureReady(ctx context.Context) error {
...
+	type WithNatIps struct {
+		VPC struct {
+			NatIPs []string `json:"natIps"`
+		} `json:"vpc"`
+	}
+	if b.Shoot.InfrastructureStatus == nil {
+		return nil, errors.New("cannot find infra status")
+	}
+	var s WithNatIps
+	err := json.Unmarshal(b.Shoot.InfrastructureStatus, s)
```
Comment (tennix) [inline pkg/operation/botanist/infrastructure.go:178]:
```suggestion
	err := json.Unmarshal(b.Shoot.InfrastructureStatus, &s)
```

File: pkg/operation/botanist/addons.go (hunk 1 above)
Comment (tennix) [inline pkg/operation/botanist/addons.go:382]: How do we create this configmap? After each seed cluster is deployed?

Comment (aylei) [reply pkg/operation/botanist/addons.go:382]: Create the configmap in pulumi script, we can get the outbound ips there

Comment (tennix) [reply pkg/operation/botanist/addons.go:382]: And does this have permission to read configmap in seed's kube-system namespace?

Comment (aylei) [reply pkg/operation/botanist/addons.go:382]: Yes, gardenlet even writes to kube-system

**Pull request merged**